api/src/flows.ts

```diff
@@ -0,0 +1,249 @@
+				}
+			} else if (flow.trigger === 'operation') {
+				const handler = (data: unknown, context: Record<string, unknown>) => this.executeFlow(flow, data, context);
+
+				this.operationFlowHandlers[flow.id] = handler;
+			} else if (flow.trigger === 'webhook' || flow.trigger === 'manual') {
+				const handler = (data: unknown, context: Record<string, unknown>) => this.executeFlow(flow, data, context);
+
+				const method = flow.options?.method ?? 'GET';
+
+				this.webhookFlowHandlers[`${method}-${flow.id}`] = handler;
+			}
+		}
+	}
+
+	public async reload(): Promise<void> {
+		for (const trigger of this.triggerHandlers) {
+			switch (trigger.type) {
+				case 'filter':
+					emitter.offFilter(trigger.name, trigger.handler);
+					break;
+				case 'action':
+					emitter.offAction(trigger.name, trigger.handler);
+					break;
+				case 'init':
+					emitter.offInit(trigger.name, trigger.handler);
+					break;
+				case 'schedule':
+					trigger.task.stop();
+					break;
+			}
+		}
+
+		this.triggerHandlers = [];
+		this.operationFlowHandlers = {};
+		this.webhookFlowHandlers = {};
+
+		this.flowOperations = {};
+
+		await this.initialize();
+	}
+
+	public addOperation(id: string, operation: OperationHandler): void {
+		this.operations[id] = operation;
+	}
+
+	public clearOperations(): void {
+		this.operations = {};
+	}
+
+	public async runOperationFlow(id: string, data: unknown, context: Record<string, unknown>): Promise<unknown> {
+		if (!(id in this.operationFlowHandlers)) {
+			logger.warn(`Couldn't find operation triggered flow with id "${id}"`);
+			return null;
+		}
+
+		const handler = this.operationFlowHandlers[id];
+
+		return handler(data, context);
+	}
+
+	public async runWebhookFlow(id: string, data: unknown, context: Record<string, unknown>): Promise<unknown> {
+		if (!(id in this.webhookFlowHandlers)) {
+			logger.warn(`Couldn't find webhook triggered flow with id "${id}"`);
+			throw new exceptions.ForbiddenException();
+		}
+
+		const handler = this.webhookFlowHandlers[id];
+
+		return handler(data, context);
+	}
+
+	public async runOperation(id: string, data: unknown): Promise<unknown> {
+		if (!(id in this.flowOperations)) {
+			logger.warn(`Couldn't find operation with id "${id}"`);
+
+			return null;
+		}
+
+		const handler = this.flowOperations[id];
+
+		return handler(data);
+	}
+
+	private async executeFlow(flow: Flow, data: unknown = null, context: Record<string, unknown> = {}): Promise<unknown> {
+		const keyedData: Record<string, unknown> = {
+			[TRIGGER_KEY]: data,
+			[LAST_KEY]: data,
+			[ACCOUNTABILITY_KEY]: context?.accountability ?? null,
+		};
+
+		let nextOperation = flow.operation;
+
+		while (nextOperation !== null) {
+			const { successor, data } = await this.executeOperation(nextOperation, keyedData, context);
+
+			keyedData[nextOperation.key] = data;
+			keyedData[LAST_KEY] = data;
+
+			nextOperation = successor;
+		}
+
+		if (flow.options.return) {
```

Comment:
Using a filter trigger currently always returns `null` for me and blocking the persistence of data. Adding something like `if (flow.trigger === 'filter') return get(keyedData, LAST_KEY);` here to simply return the last object could fix it quickly but having something like an "output" arrow specifically for filters might be a nicer solution in the long run.

Reply:
That's exactly what `flow.options.return` is for. This should be configurable in the filter trigger's options. We should probably set this to `$last` or `$trigger` by default to at least return something.

Adding an "output" arrow might be tricky because arrows indicate control flow and not necessarily data flow.

Reply:
I had a suspicion thats what it would be for but wasnt sure without context. An arrow is indeed overkill if this can be a select on the trigger with a sane default 😄 that sounds like something i could implement too

Reply:
`executeFlow()` would previously return `undefined` if `flow.options.return` was not defined. This would let the event emitter use the previous payload without changing it.

@rijkvanzanten were you running into issues with returning `undefined`?

Reply:
@nickrum There was an issue where the webhook trigger would return string `null` as the output payload by default, I believe my change was to combat that